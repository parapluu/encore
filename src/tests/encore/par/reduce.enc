def sum(acc: int, y: int): int
  acc + y

def millionItems(size: int): Par int {
  var p = liftv 0;
  repeat i <- size {
    p = p || liftv(i);
  };
  p
}

def millionFutures(size: int): Par int {
  val actors = initActors(size);
  execute(actors)
} where
  def initActors(size: int): [Accountant] {
    val ar = new [Accountant](size);
    repeat i <- size {
      ar[i] = new Accountant(i);
    };
    ar
  }
  def execute(actors: [Accountant]): Par int {
    var p = liftv 0;
    for a in actors {
      p = p || liftf(a.getNumber());
    };
    p
  }
end


class Accountant
  number: int

  def init(number: int): void
    this.number = number

  def getNumber(): int
    this.number


class Main
  def main(): void
    let f = reduce(sum, 0, millionItems(1000000))
        f2 = reduce(sum, 0, millionFutures(1000000))
        v = get f
        v2 = get f2
    in {
      print("Result million values: {}\n", v);
      print("Result million futures: {}\n", v2);
    }
