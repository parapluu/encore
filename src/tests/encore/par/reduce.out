Result million values: 499999500000
Result million futures: 499999500000
